sdk/include/platform/sunburst/platform-ethernet.hh

```diff
@@ -0,0 +1,633 @@
+	{
+		uint8_t addr = static_cast<uint8_t>(reg);
+		uint8_t be   = (addr & 0x2) == 0 ? 0b0011 : 0b1100;
+		uint8_t bytes[2];
+		bytes[0] = (0b01 << 6) | (be << 2) | (addr >> 6);
+		bytes[1] = (addr << 2) & 0b11110000;
+
+		set_gpio_output_bit(EthCsPin, 0);
+		spi()->blocking_write(bytes, 2);
+		spi()->blocking_write((uint8_t *)&val, 2);
+		spi()->wait_idle();
+		set_gpio_output_bit(EthCsPin, 1);
+	}
+
+	void register_set(RegisterOffset reg, uint16_t mask) const
+	{
+		uint16_t old = register_read(reg);
+		register_write(reg, old | mask);
+	}
+
+	void register_clear(RegisterOffset reg, uint16_t mask) const
+	{
+		uint16_t old = register_read(reg);
+		register_write(reg, old & ~mask);
+	}
+
+	/**
+	 * Helper.  Returns a pointer to the SPI device.
+	 */
+	[[nodiscard, gnu::always_inline]] Capability<volatile SonataSpi> spi() const
+	{
+		return MMIO_CAPABILITY(SonataSpi, spi2);
+	}
+
+	/**
+	 * Helper.  Returns a pointer to the GPIO device.
+	 */
+	[[nodiscard, gnu::always_inline]] Capability<volatile SonataGPIO>
+	gpio() const
+	{
+		return MMIO_CAPABILITY(SonataGPIO, gpio);
+	}
+
+	uint16_t frames_to_process = 0;
+
+	public:
+	/**
+	 * Initialise a reference to the Ethernet device.
+	 */
+	Ksz8851Ethernet()
+	{
+		// Reset chip
+		set_gpio_output_bit(EthRstPin, false);
+		thread_millisecond_wait(150);
+		set_gpio_output_bit(EthRstPin, true);
+
+		uint16_t cider = register_read(RegisterOffset::CIDER);
+		Debug::log("Chip ID is {}", cider);
+
+		// Check the chip ID. The last nibble is revision ID and can be ignored.
+		Debug::Assert((cider & 0xFFF0) == 0x8870, "Unexpected Chip ID");
+
+		// Enable QMU Transmit Frame Data Pointer Auto Increment
+		register_write(RegisterOffset::TXFDPR, 0x4000);
+		// Enable QMU Transmit flow control / Transmit padding / Transmit CRC,
+		// and IP/TCP/UDP checksum generation.
+		register_write(RegisterOffset::TXCR, 0x00EE);
+		// Enable QMU Receive Frame Data Pointer Auto Increment.
+		register_write(RegisterOffset::RXFDPR, 0x4000);
+		// Configure Receive Frame Threshold for one frame.
+		register_write(RegisterOffset::RXFCTR, 0x0001);
+		// Enable QMU Receive flow control / Receive all broadcast frames
+		// /Receive unicast frames, and IP/TCP/UDP checksum verification etc.
+		register_write(RegisterOffset::RXCR1, 0x7CE0);
+		// Enable QMU Receive UDP Lite frame checksum verification, UDP Lite
+		// frame checksum generation, IPv6 UDP fragment frame pass, and
+		// IPv4/IPv6 UDP UDP checksum field is zero pass. In addition (not in
+		// the programmer's guide), enable single-frame data burst.
+		register_write(RegisterOffset::RXCR2, 0x009C);
+		// Enable QMU Receive Frame Count Threshold / RXQ Auto-Dequeue frame.
+		register_write(RegisterOffset::RXQCR,
+		               RxQueueControl::RxFrameCountThresholdEnable |
+		                 RxQueueControl::AutoDequeueRxQFrameEnable);
+
+		// Programmer's guide have a step to set the chip in half-duplex when
+		// negotiation failed, but we omit the step.
```

Comment:
I think it's not useful because people don't use non-switching hubs and half-duplex Ethernet anymore?

Reply:
Please say that in the comment.  I think, for the A7, we force the Ethernet adaptor into 10 Mb full-duplex mode, because any switch will be able to handle that and we get to use buffering on the switch instead of dropping frames if someone tries sending messages too fast for us.